src/Kestrel.Core/Internal/ConnectionLimitMiddleware.cs

```diff
@@ -1,32 +1,74 @@
+
+using System.Threading.Tasks;
 using Microsoft.AspNetCore.Protocols;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
 {
     public class ConnectionLimitMiddleware
     {
-        private readonly ServiceContext _serviceContext;
         private readonly ConnectionDelegate _next;
+        private readonly ResourceCounter _concurrentConnectionCounter;
+        private readonly IKestrelTrace _trace;
 
-        public ConnectionLimitMiddleware(ConnectionDelegate next, ServiceContext serviceContext)
+        public ConnectionLimitMiddleware(ConnectionDelegate next, long connectionLimit, IKestrelTrace trace)
+            : this(next, ResourceCounter.Quota(connectionLimit), trace)
+        {
+        }
+
+        // For Testing
+        internal ConnectionLimitMiddleware(ConnectionDelegate next, ResourceCounter concurrentConnectionCounter, IKestrelTrace trace)
         {
             _next = next;
-            _serviceContext = serviceContext;
+            _concurrentConnectionCounter = concurrentConnectionCounter;
+            _trace = trace;
         }
 
-        public Task OnConnectionAsync(ConnectionContext connection)
+        public async Task OnConnectionAsync(ConnectionContext connection)
         {
-            if (!_serviceContext.ConnectionManager.NormalConnectionCount.TryLockOne())
+            if (!_concurrentConnectionCounter.TryLockOne())
             {
                 KestrelEventSource.Log.ConnectionRejected(connection.ConnectionId);
-                _serviceContext.Log.ConnectionRejected(connection.ConnectionId);
+                _trace?.ConnectionRejected(connection.ConnectionId);
                 connection.Transport.Input.Complete();
                 connection.Transport.Output.Complete();
-                return Task.CompletedTask;
+                return;
+            }
+
+            var releasor = new ConnectionReleasor(_concurrentConnectionCounter);
+
+            try
+            {
+                connection.Features[typeof(IDecrementConcurrentConnectionCountFeature)] = releasor;
+                await _next(connection);
+            }
+            finally
+            {
+                releasor.ReleaseConnection();
             }
+        }
 
-            return _next(connection);
+        private class ConnectionReleasor : IDecrementConcurrentConnectionCountFeature
+        {
+            private readonly ResourceCounter _concurrentConnectionCounter;
+            private bool _connectionReleased;
+
+            public ConnectionReleasor(ResourceCounter normalConnectionCounter)
+            {
+                _concurrentConnectionCounter = normalConnectionCounter;
+            }
+
+            public void ReleaseConnection()
+            {
+                if (!_connectionReleased)
```

Comment:
Does this need to be thread safe?

Reply:
I almost commented on the interface that this isn't thread safe, but I figure that was expected. Most connection/request features like the header dictionaries aren't thread safe.